packages/database/src/core/BaseDatabase.js

```diff
@@ -460,6 +456,100 @@ export class BaseDatabase {
   get isWithinTransaction() {
     return this.connection.isTransaction;
   }
+
+  createChangeChannel() {
+    throw new Error('createChangeChannel should be implemented by the child class');
+  }
+
+  getHandlersForChange(change) {
+    const { handler_key: specificHandlerKey, record_type: recordType } = change;
+    const handlersForCollection = this.getChangeHandlersForCollection(recordType);
+    if (specificHandlerKey) {
+      return handlersForCollection[specificHandlerKey]
+        ? [handlersForCollection[specificHandlerKey]]
+        : [];
+    }
+    return Object.values(handlersForCollection);
+  }
+
+  async notifyChangeHandlers(change) {
+    const unlock = this.handlerLock.createLock(change.record_id);
+    const handlers = this.getHandlersForChange(change);
+    const scheduledPromises = [];
+    try {
+      for (const handler of handlers) {
+        try {
+          const { scheduledPromise } = (await handler(change)) || {};
+          if (scheduledPromise) {
+            scheduledPromises.push(scheduledPromise);
+          }
+        } catch (e) {
+          winston.error(e);
+        }
+      }
+    } finally {
+      // Don't await the scheduled promises, so that we don't block the change handler from completing
+      Promise.all(scheduledPromises).finally(unlock);
+    }
+  }
+
+  async waitForAllChangeHandlers() {
+    return this.handlerLock.waitWithDebounce(HANDLER_DEBOUNCE_DURATION);
+  }
+
+  getChangeHandlersForCollection(collectionName) {
+    // Instantiate the array if no change handlers currently exist for the collection
+    if (!this.changeHandlers[collectionName]) {
+      this.changeHandlers[collectionName] = {};
+    }
```

Comment:
This is a fun one:

```suggestion
    this.changeHandlers[collectionName] ??= {};
```

And because an assignment statement returns the new value, you could even inline it

```ts
return this.changeHandlers[collectionName] ??= {};
```

Reply:
really handy!